src/Open3D/GUI/SceneWidget.cpp

```diff
@@ -124,11 +144,321 @@ class CameraControls {
+        // would be incorrect, since GetForwardVector() returns the forward
+        // vector in world space, but the translation happens in camera space.)
+        // Since we want trackpad down (negative) to go forward ("pulling" the
+        // model toward the viewer) we need to negate dy.
+        auto forward = Eigen::Vector3f(0, 0, -zDist);  // zDist * (0, 0, -1)
+        matrix.translate(forward);
+        matrix_ = matrix;
+    }
+
+private:
+    visualization::Camera::Transform matrix_;
+
+    visualization::Camera::Transform matrixAtMouseDown_;
+    Eigen::Vector3f centerOfRotationAtMouseDown_;
+
+protected:
+    Size viewSize_;
+    double modelSize_ = 100.0;
+    geometry::AxisAlignedBoundingBox modelBounds_;
+    Eigen::Vector3f centerOfRotation_;
+};
+
+class LightDirControl : public MatrixControl {
+    using Super = MatrixControl;
+public:
+    LightDirControl(visualization::Camera *camera)
+    : camera_(camera) {
+    }
+
+    void SetDirectionalLight(visualization::Scene *scene,
+                             visualization::LightHandle dirLight) {
+        scene_ = scene;
+        dirLight_ = dirLight;
+    }
+
+    void Rotate(int dx, int dy) override {
+        Eigen::Vector3f up = camera_->GetUpVector();
+        Eigen::Vector3f right = -camera_->GetLeftVector();
+        RotateWorld(-dx, -dy, up, right);
+        UpdateMouseDragUI();
+    }
+
+    void StartMouseDrag() {
+        lightDirAtMouseDown_ = scene_->GetLightDirection(dirLight_);
+        auto identity = visualization::Camera::Transform::Identity();
+        Super::SetMouseDownInfo(identity, {0.0f, 0.0f, 0.0f});
+
+        for (auto &o : uiObjs_) {
+            scene_->RemoveGeometry(o.handle);
+        }
+
+        auto dir = scene_->GetLightDirection(dirLight_);
+
+        // TODO: it seems that SetEntityTransform moves an object to be
+        //       centered about the origin before applying the transform.
+        //       This appears to be a Filament behavior, which we work around
+        //       here. This code needs to be changed once we fix that
+        //       behavior.
+
+        double sphereSize = 0.5 * modelSize_; // modelSize_ is a diameter
+        auto sphereTris = geometry::TriangleMesh::CreateSphere(sphereSize, 20);
+        auto sphere = geometry::LineSet::CreateFromTriangleMesh(*sphereTris);
+        sphere->PaintUniformColor({0.0f, 0.0f, 1.0f});
+        auto t0 = visualization::Camera::Transform::Identity();
+        uiObjs_.push_back({scene_->AddGeometry(*sphere), t0});
+        scene_->SetEntityTransform(uiObjs_[0].handle, t0);
+
+        auto sunRadius = 0.05 * modelSize_;
+        auto sun = geometry::TriangleMesh::CreateSphere(sunRadius, 20);
+        sun->PaintUniformColor({1.0f, 0.5f, 0.0f});
+        auto t1 = visualization::Camera::Transform::Identity();
+        t1.translate(-sphereSize * dir);
+        uiObjs_.push_back({scene_->AddGeometry(*sun), t1});
```

Comment:
I was going to leave it as a bugfix, and talk to you about the best way to update the Scene::AddGeometry, since there are a three of them. Do I add a useShadows=true parameter to all three? Do you have other things that need to be set, so we should use something like LightDescriptor for lights? Or should I add a SetGeometryShadows() function?

Reply:
`SetGeometryShadows` suits better for current API